cpmpy/transformations/linearize.py

```diff
@@ -1,56 +1,63 @@
-
-Linear comparison:
-------------------
-- ``LinExpr == Constant``
-- ``LinExpr >= Constant``
-- ``LinExpr <= Constant``
-
-LinExpr can be any of:
-
-- `NumVar`
-- `sum`
-- `wsum`
-
-Indicator constraints:
-----------------------
-
-+------------------------------------+
-| ``BoolVar -> LinExpr == Constant`` |
-+------------------------------------+
-| ``BoolVar -> LinExpr >= Constant`` |
-+------------------------------------+
-| ``BoolVar -> LinExpr <= Constant`` |
-+------------------------------------+
-
-========================================   ==============================================
-``BoolVar -> GenExpr``                     (GenExpr.name in supported, GenExpr.is_bool()) 
-``BoolVar -> GenExpr >= Var/Constant``     (GenExpr.name in supported, GenExpr.is_num())  
-``BoolVar -> GenExpr <= Var/Constant``     (GenExpr.name in supported, GenExpr.is_num())  
-``BoolVar -> GenExpr == Var/Constant``     (GenExpr.name in supported, GenExpr.is_num())  
-========================================   ==============================================
-
-Where ``BoolVar`` is a boolean variable or its negation.
-
-General comparisons or expressions
------------------------------------
-
-============================  ==============================================
-``GenExpr``                   (GenExpr.name in supported, GenExpr.is_bool())  
-``GenExpr == Var/Constant``   (GenExpr.name in supported, GenExpr.is_num())  
-``GenExpr <= Var/Constant``   (GenExpr.name in supported, GenExpr.is_num())  
-``GenExpr >= Var/Constant``   (GenExpr.name in supported, GenExpr.is_num()) 
-============================  ============================================== 
-
-
-
+Transform constraints to a linear form.
+
+This transformation is necessary for Integer Linear Programming (ILP) solvers, and also
+for translating to Pseudo-Boolean or CNF formats.
+
+There are a number of components to getting a good linearisation:
+- **Decomposing global constraints/functions** in a 'linear friendly' way.
+  A common pattern is that constraints that enforce domain consistency on integer variables,
+  that they should be decomposed over a Boolean representation of the domain. This is the
+  case for :class:`~cpmpy.expressions.globalconstraints.AllDifferent`, :class:`~cpmpy.expressions.globalfunctions.Element`
+  and possibly others.
+  Their default decomposition might not do it this way, in which case we want to use different
+  decompositions.
+
+- **Linearising multiplication** of variables, e.g. bool*bool, bool*int and int*int.
+  Because '*' is a core Operator and not a global constraint, we need to do it with a helper function here.
+
+- **Disequalities** e.g. `sum(X) != 5` should be rewritten as `(sum(X) < 5) | (sum(X) > 5)`
+  and further flattened into implications and linearised.
+
+- **Implications** e.g. `B -> sum(X) <= 4` should be linearised with the Big-M method.
+
+- **Domain encodings** e.g. of `A=cp.intvar(0,4)` would create binary variables and constraints
```

Comment:
This is a bit surprising to see in the docs here for me.
The other points are "constraint xxx is not linear, and should be linearized as follows...."; but here there is no "input" constraint?